bip-unknown-var-budget-script.mediawiki

```diff
@@ -0,0 +1,722 @@
+# If there was final overflow, fail.
+# Remember last non-zero byte written into A, and truncate A after that point.
+|-
+|OP_1SUB
+|140	
+|1
+|
+# Pop A off the stack.
+# Let B = 1, and continue as OP_SUB.
+|}
+
+====Rationale====
+
+These operations may seem overspecified, but our cost model requires that we only make a single pass over each parameter (with the exception of appending a final carry for add).  This is why each is defined to be done in-place rather than using a separate result vector.  Also, a naive implementation of add or subtract might simply trim trailing zeroes as a final step, which could violate our cost model.  For add this can be avoided by either pre-trimming the operands (option 1), or post-trimming the result (option 2).
+
+Subtraction underflow is illegal: mathematicians agree the result would not be natural.
+
+===Variable Opcode Budget===
+
+Each opcode in v2 tapscript has a varops budget cost, based on the length of data processed.
+
+- COMPARING: Comparing two objects: cost = 1 per byte compared.
+- COMPARINGZERO: Comparing an object against zeroes: cost = 1 per byte compared.
+- ZEROING: Zeroing out bytes: cost = 1 per byte zeroed
+- COPYING: Copying bytes: cost = 1 per byte copied.
+- All other operations: cost = 2 per byte.
+
+In practice, this mirrors the measured speed of implementations on a modern CPU quite well.  Note that comparing an array against zeros can be done by comparing the first few bytes then an offset comparison with itself.
+
+Some operations have corner cases which require more than one pass over the operands, and these are defined carefully below.  No opcode currently accesses randomly within stack objects: such an opcode would require an extension of this cost model.
+
+Opcodes which neither examine nor produce variable-length stack objects are assigned cost 0, and are not mentioned here.
+
+We do not consider the cost of the script interpretation itself, which is necessarily limited by block size.
+
+There are various places where script interprets a value as a length.  In practice these all have a well-known upper bound, but the operand (potentially 4MB worth) has to be examined past the first 64 bytes to ensure the remainder is zero.  These are annotated as "LENGTHCONV" and costed as a COMPARINGZERO operation.
+
+===Control And Simple Examination Opcodes===
+
+{|
+! Opcode
+! Varops Budget Cost
+|-
+|OP_VERIFY
+|Operand length (COMPARINGZERO)
+|-
+|OP_NOT
+|Operand length (COMPARINGZERO)
+|-
+|OP_0NOTEQUAL
+|Operand length (COMPARINGZERO)
+|-
+|OP_EQUAL
+|If length unequal: 0, otherwise length (COMPARING)
+|-
+|OP_EQUALVERIFY
+|If length unequal: 0, otherwise length (COMPARING)
+|}
+
+====Rationale====
+
+In the worst case, an implementation must examine every byte of a stack element to determine if the value is 0.  This can be done efficiently using existing comparison techniques, by examining the first byte, and then using a offet comparison against itself.
```

Comment:
```suggestion
In the worst case, an implementation must examine every byte of a stack element to determine if the value is 0.  This can be done efficiently using existing comparison techniques, by examining the first byte, and then using an offset comparison against itself.
```